src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchy.cs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace NuGet.VisualStudio
+{
+    /// <summary> VS hierarchy helper. </summary>
+    public sealed class VsHierarchy
+    {
+        private static readonly string[] UnsupportedProjectCapabilities = new string[]
+        {
+            "SharedAssetsProject", // This is true for shared projects in universal apps
+        };
+
+        private readonly IVsHierarchy _hierarchy;
+
+        private VsHierarchy(IVsHierarchy hierarchy)
+        {
+            _hierarchy = hierarchy;
+        }
+
+        /// <summary> Create a new instance of <see cref="VisualStudio.VsHierarchy"/> from a DTE project object. </summary>
+        /// <param name="project"> A DTE project. </param>
+        /// <returns> Instance of <see cref="VisualStudio.VsHierarchy"/> wrapping <paramref name="project"/>. </returns>
+        public static async Task<VsHierarchy> FromDteProjectAsync(EnvDTE.Project project)
+        {
+            Assumes.Present(project);
+            return new VsHierarchy(await ToVsHierarchyAsync(project));
+        }
+
+        /// <summary> Create a new instance of <see cref="VisualStudio.VsHierarchy"/> from a <see cref="IVsHierarchy"/> object. </summary>
+        /// <param name="project"> A <see cref="IVsHierarchy"/> project. </param>
+        /// <returns> Instance of <see cref="VisualStudio.VsHierarchy"/> wrapping <paramref name="project"/>. </returns>
+        public static VsHierarchy FromVsHierarchy(IVsHierarchy project)
+        {
+            Assumes.Present(project);
+            return new VsHierarchy(project);
+        }
+
+        /// <summary> Underlying <see cref="IVsHierarchy"/> object. </summary>
+        public IVsHierarchy Ptr => _hierarchy;
```

Comment:
I would assert here for UI thread to make it very clear that this "Ptr" cannot be used off UI thread. The async'ing of these methods makes it very easy to do something like this which is not obvious to spot in code review:

``` C#
var hierarchy = await GetSomeThingAsync();

// This will RPC in legacy project system and block a thread while it waits for UI thread to respond
return (IVsUIHierarchy)hierarchy.Ptr;

```

``` C#
var project = await GetDteProjectAsync();

// Blocking RPC for every loop
foreach (ProjectItem item in project.ProjectItems)
{
     // Another blocking RPC here
    if (item.Bar)
    {
         // Another blocking RPC here
         item.Foo();
    }
}
```

You should make it impossible to do the above patterns.

Reply:
> I would assert here for UI thread to make it very clear that this "Ptr" cannot be used off UI thread.

Wouldn't that mess with the scenarios that access this `IVsHierarchy` to pass it through?
It wouldn't need the UI thread otherwise.

Reply:
They should be passing this handle to it, if not, this PR makes me worried as it makes it much easier to do the wrong thing if you return IVsHierarchy asynchronously back to background threads.

Reply:
I intended to have `Ptr` as a temporary measure until I can track down its usage and actually wrap all _actual_ operations needing UI thread into methods on this wrapper.

Reply:
Here's our equivalent: https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsUIService%601.cs